pkg/cloudprovider/mesos/client.go

```diff
@@ -0,0 +1,296 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mesos
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/resource"
+	log "github.com/golang/glog"
+	"github.com/mesos/mesos-go/detector"
+	mesos "github.com/mesos/mesos-go/mesosproto"
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultClusterName = "mesos"
+)
+
+var (
+	noLeadingMasterError = fmt.Errorf("there is no current leading master available to query")
+)
+
+type mesosClient struct {
+	masterLock    sync.RWMutex
+	master        string // host:port formatted address
+	httpClient    *http.Client
+	tr            *http.Transport
+	initialMaster <-chan struct{} // signal chan, closes once an initial, non-nil master is found
+	state         *stateCache
+}
+
+type slaveNode struct {
+	hostname  string
+	resources *api.NodeResources
+}
+
+type mesosState struct {
+	clusterName string
+	nodes       []*slaveNode
+}
+
+type stateCache struct {
+	sync.Mutex
+	expiresAt time.Time
+	cached    *mesosState
+	err       error
+	ttl       time.Duration
+	refill    func(context.Context) (*mesosState, error)
+}
+
+// Reload the state cache if it has expired.
+func (c *stateCache) reloadCache(ctx context.Context) {
+	now := time.Now()
+	c.Lock()
+	defer c.Unlock()
+	if c.expiresAt.Before(now) {
+		log.V(4).Infof("Reloading cached Mesos state")
+		c.cached, c.err = c.refill(ctx)
+		c.expiresAt = now.Add(c.ttl)
+	} else {
+		log.V(4).Infof("Using cached Mesos state")
+	}
+}
+
+// Returns the cached Mesos state.
+func (c *stateCache) getCachedState(ctx context.Context) (*mesosState, error) {
+	c.reloadCache(ctx)
+	return c.cached, c.err
+}
+
+// Returns the cached Mesos cluster name.
+func (c *stateCache) getClusterName(ctx context.Context) (string, error) {
```

Comment:
nit: "getter" functions in go don't start with "get". This should be renamed to "clusterName"